product_customer_code/models/product.py

```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+
+###########################################################################
+#    Module Writen to OpenERP, Open Source Management Solution
+#
+#    Copyright (c) 2012 Vauxoo - http://www.vauxoo.com
+#    All Rights Reserved.
+#    info@vauxoo.com
+############################################################################
+#    Coded by: Rodo (rodo@vauxoo.com),Moy (moylop260@vauxoo.com)
+############################################################################
+
+from odoo import models, fields, api
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    product_customer_code_ids = fields.One2many(
+        comodel_name='product.customer.code',
+        inverse_name='product_id',
+        string='Customer Codes',
+        copy=False
+    )
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
+        res = super(ProductProduct, self).name_search(
+            name=name, args=args, operator=operator, limit=limit)
+        if not res:
+            partner_id = self._context.get('partner_id')
+            if partner_id:
+                product_customer_code_obj = self.env['product.customer.code']
+                product_codes = product_customer_code_obj.search([
+                    ('product_code', '=', name),
```

Comment:
@lasley we have the identical use case, except mine is real.  We deal with many national entities - think McDonalds, where there are literally hundreds of sub-entities that all use the exact same product code but are different entities.  However I think the entire argument can be alleviated to cover both use cases simply by
```
'|', ('partner_id', '=' partner.id),('partner_id','=', partner.commercial_id.id)
```
Then update README with expected behaviour.  That actually doesn't get my use case, because we also bill these sub-entities but we wouldn't use this module either.

I would need (pseudocode)
```
('partner_id', 'in' [partner.id, partner.commercial_partner_id.id, partner.parent_id.id,partner.commercial_partner_id.parent_id.id])
```
which I think is also reasonable.  In practice, its unlikely for any given customer entity that they would mix behaviours and this covers where product codes are set at a higher level.  We get individual, entity, parent entity in even reasonably complex setups.

I agree searching children is wrong, was just typing out loud at the time.